kitsune/sumo/api_utils.py

```diff
@@ -298,51 +253,35 @@ def has_object_permission(self, request, view, obj):
         return user == creator
 
 
-def PermissionMod(cls, permissions):
+PermissionListSerializer = None
+
+
+def PermissionMod(field, permissions=[]):
     """
     Takes a class and modifies it to conditionally hide based on permissions.
     """
-    class Modded(cls):
-
-        def __init__(self, **kwargs):
-            super(Modded, self).__init__(**kwargs)
-            self._stealth = False
-
-        def field_to_native(self, obj, field_name):
-            """
-            Return null value if request has no access to that field
-            """
-            if self.check_permissions() and self.check_object_permissions(obj):
-                self._stealth = False
-                return super(Modded, self).field_to_native(obj, field_name)
-            self._stealth = True
-            return None
-
-        # write_only is converted to a property, so that it can be "stealthed."
-        # The write only field is how DRF chooses to include or not include a field
-        # in serialization, so by controlling it, this field can choose when it
-        # is included in serialization output.
 
-        @property
-        def write_only(self):
-            return self._stealth or self._write_only
+    class Modded(field):
+        @classmethod
+        def many_init(cls, *args, **kwargs):
+            kwargs['child'] = field()
+            return PermissionMod(serializers.ListSerializer, permissions)(*args, **kwargs)
 
-        @write_only.setter
-        def write_only(self, value):
-            self._write_only = value
-
-        # These methods mimic the methods found on DRF views.
-
-        def get_permissions(self):
-            return (p() for p in permissions)
-
-        def check_permissions(self):
-            request = self.context.get('request', None)
-            return all(p.has_permission(request, self) for p in self.get_permissions())
-
-        def check_object_permissions(self, obj):
-            request = self.context.get('request', None)
-            return all(p.has_object_permission(request, self, obj) for p in self.get_permissions())
+        def get_attribute(self, instance):
+            if self.check_permissions(instance):
+                return super(Modded, self).get_attribute(instance)
+            else:
+                raise fields.SkipField()
+
+        def check_permissions(self, obj):
+            request = self.context.get('request')
+            for Perm in permissions:
+                perm = Perm()
+                if not perm.has_permission(request, self):
+                    return False
+                if not perm.has_object_permission(request, self, obj):
+                    return False
+            return True
 
     return Modded
```

Comment:
I think that the Stack post you linked is talking about something else. They seem to be talking about the way of writing decorators where you have a class with an `__init__` and a `__call__` method that returns a function, instead of having triply nested functions.

I'm not sure this matters. When debugging decorated functions, there isn't any standard way to get at the original function, so you try really hard to make sure not to erase information. With a class decorator that works by subclassing like this, you can get back the original class (`decorated_class.__class__.__mro__[0]`). It isn't really easy, but it does exist.

We've had this decorator-a-like for a while now, and it hasn't been a problem that we aren't doing any metadata copying on it. Further, I did a bunch of debugging around it while working on this patch, and I didn't ever think "damn, I wish I had just preserved the metadata".

Given those things, I think I'm going to leave this like it, unless you have any clever ideas on how to copy the metadata we might want.